
The points distribution for this case is as follows:
1. Read the dataset in a new python notebook.
2. Drop the columns which are unique for all users like IDs (2.5 points)
3. Distinguish the feature and target set (2.5 points)
4. Divide the data set into Train and test sets
5. Normalize the train and test data (2.5 points)
6. Initialize &amp; build the model (10 points)
7. Optimize the model (5 points)
9. Predict the results using 0.5 as a threshold (5 points)
10. Print the Accuracy score and confusion matrix (2.5 points)
 


### Given a Bank customer, can we build a classifier which can determine whether they will leave or not using Neural networks?


In [1]:
import tensorflow as tf
print(tf.__version__)

C:\Users\DELL\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\Users\DELL\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\Users\DELL\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\Users\DELL\Anaconda3\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:5

2.0.0


In [2]:
import pandas as pd
import numpy as np

In [3]:
#Read the dataset in a new python notebook
df=pd.read_csv("bank.csv")

In [4]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


### Drop the columns which are unique for all users like IDs

In [6]:

df=df.drop(columns=['RowNumber','CustomerId','Surname'])

- Looking at the features we can see that row number, customer ID,surname will have no relation with a customer with leaving the bank.

In [7]:
df.columns

Index(['CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance',
       'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary',
       'Exited'],
      dtype='object')

In [8]:
from sklearn.preprocessing import LabelEncoder
categorical_classes_list = ['Geography','Gender'] 
#encode features that are cateorical classes
encoding_list = []
for column in categorical_classes_list:
    le = LabelEncoder()
    le.fit(df[column])
    encoding_list.append(df[column].unique())
    encoding_list.append(list(le.transform(df[column].unique())))
    df[column] = le.transform(df[column])
    #test_set[column] = le.transform(test_set[column])

### Distinguish the feature and target set 

In [9]:
X = df.drop(columns=['Exited'])
Y = df['Exited']

### Divide the data set into Train and test sets

In [10]:
from sklearn.model_selection import train_test_split
train_x,test_x,train_y,test_y=train_test_split(X,Y,test_size=.30,random_state=2)

In [11]:
train_x =np.array(train_x).astype('float32')
test_x = np.array(test_x).astype('float32')
train_y =np.array(train_y).astype('float32')
test_y = np.array(test_y).astype('float32')

### Normalize the train and test data

In [12]:
from sklearn.preprocessing import Normalizer

transformer = Normalizer()
train_x = transformer.fit_transform(train_x)
test_x = transformer.transform(test_x)

In [13]:
train_x.shape

(7000, 10)

In [14]:
test_x.shape

(3000, 10)

In [15]:
train_y.shape

(7000,)

In [16]:
train_y=train_y.reshape(-1,1)

In [17]:
train_y.shape

(7000, 1)

### Initialize & build the model

In [18]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import regularizers
from tensorflow.keras import optimizers
from tensorflow.keras.backend import backend

In [19]:
#Adding the input layer and the first hidden layer…
classifier = Sequential()
classifier.add(Dense(10, activation = 'relu', input_dim = 10))


In [20]:
#Adding the second hidden layer…
classifier.add(Dense(10, activation = 'relu'))

In [21]:
#Adding the output layer…
classifier.add(Dense(1, activation = 'sigmoid'))

### Optimize the model 

In [22]:
#Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [23]:
classifier.fit(train_x, train_y, validation_data=(test_x, test_y), epochs=30,batch_size = 32)

Train on 7000 samples, validate on 3000 samples
Epoch 1/30
7000/7000 [==============================] - 1s 139us/sample - loss: 0.5724 - accuracy: 0.7460 - val_loss: 0.4949 - val_accuracy: 0.8050
Epoch 2/30
7000/7000 [==============================] - 0s 50us/sample - loss: 0.5071 - accuracy: 0.7926 - val_loss: 0.4893 - val_accuracy: 0.8050
Epoch 3/30
7000/7000 [==============================] - 0s 50us/sample - loss: 0.5038 - accuracy: 0.7926 - val_loss: 0.4895 - val_accuracy: 0.8050
Epoch 4/30
7000/7000 [==============================] - 0s 49us/sample - loss: 0.5034 - accuracy: 0.7926 - val_loss: 0.4876 - val_accuracy: 0.8050
Epoch 5/30
7000/7000 [==============================] - 0s 49us/sample - loss: 0.5033 - accuracy: 0.7926 - val_loss: 0.4872 - val_accuracy: 0.8050
Epoch 6/30
7000/7000 [==============================] - 0s 50us/sample - loss: 0.5030 - accuracy: 0.7926 - val_loss: 0.4870 - val_accuracy: 0.8050
Epoch 7/30
7000/7000 [==============================] - 0s 50us/sampl

### Predict the results using 0.5 as a threshold

In [24]:
# Predicting the Test set results

y_pred = classifier.predict(test_x)
y_pred = (y_pred > 0.5)

- if y_pred is larger than 0.5 it returns true(1) else false(2)

In [25]:
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

### Print the Accuracy score and confusion matrix 

In [26]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(test_y, y_pred)
print(cm)

[[2415    0]
 [ 585    0]]


- Out of 3000 customers 2415 were predicted accurately and 585 customers were predicted inaccurately.


In [27]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(test_y, y_pred)
print('Accuracy: %f' % accuracy)

Accuracy: 0.805000
